compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -1381,25 +1379,24 @@ impl DepNodeColorMap {
     }
 }
 
-fn backtrace_printer<'a, K: DepKind>(
-    sess: &'a rustc_session::Session,
-    graph: &'a DepGraphData<K>,
-    node: SerializedDepNodeIndex,
-) -> OnDrop<impl Fn() + 'a> {
-    OnDrop(
-        #[inline(never)]
-        #[cold]
-        move || {
-            let node = graph.previous.index_to_node(node);
-            // Do not try to rely on DepNode's Debug implementation, since it may panic.
-            let diag = rustc_errors::Diagnostic::new(
-                rustc_errors::Level::FailureNote,
-                &format!(
-                    "encountered while trying to mark dependency green: {:?}({})",
-                    node.kind, node.hash
-                ),
-            );
-            sess.diagnostic().force_print_diagnostic(diag);
-        },
-    )
+#[inline(never)]
+#[cold]
+pub(crate) fn print_markframe_trace<K: DepKind>(
+    graph: &DepGraphData<K>,
+    frame: Option<&MarkFrame<'_>>,
+) {
+    eprintln!("there was a panic while trying to force a dep node");
+    eprintln!("try_mark_green dep node stack:");
+
+    let mut i = 0;
+    let mut current = frame;
+    while let Some(frame) = current {
+        // Do not try to rely on DepNode's Debug implementation, since it may panic.
+        let node = graph.previous.index_to_node(frame.index);
+        eprintln!("#{i} {:?} ({})", node.kind, node.hash);
```

Comment:
Should we try to show the debug version of `node`? That'd be more helpful with the pretty name of a DefId instead of its hash.

```suggestion
        eprintln!("#{} [{:?}] {:?}", i, node.kind, node.hash);
```

Reply:
Yeah, that's probably worth the panic risk. Unwinding happens after the ICE message anyway.